client-core/src/transaction_builder/default_transaction_builder.rs

```diff
@@ -0,0 +1,441 @@
+use chain_core::tx::TxAux;
+use client_common::{ErrorKind, Result};
+
+use crate::{TransactionBuilder, WalletClient};
+
+/// Default implementation of `TransactionBuilder`
+pub struct DefaultTransactionBuilder<F>
+where
+    F: FeeAlgorithm,
+{
+    fee_algorithm: F,
+}
+
+impl<F> DefaultTransactionBuilder<F>
+where
+    F: FeeAlgorithm,
+{
+    /// Creates a new instance of `DefaultTransactionBuilder`
+    pub fn new(fee_algorithm: F) -> Self {
+        Self { fee_algorithm }
+    }
+
+    fn amount_to_transfer(&self, outputs: &[TxOut], fee: Fee) -> Result<Coin> {
+        let mut amount_to_transfer = fee.to_coin();
+        for output in outputs.iter() {
+            amount_to_transfer =
+                (amount_to_transfer + output.value).context(ErrorKind::BalanceAdditionError)?;
+        }
+
+        Ok(amount_to_transfer)
+    }
+
+    fn select_transactions(
+        &self,
+        mut unspent_transactions: Vec<(TxoPointer, Coin)>,
+        amount_to_transfer: Coin,
+    ) -> Result<(Vec<TxoPointer>, Coin)> {
+        unspent_transactions.sort_by(|a, b| a.1.cmp(&b.1).reverse());
+
+        let mut selected_unspent_transactions = Vec::new();
+        let mut transferred_amount = Coin::zero();
+        for (unspent_transaction, value) in unspent_transactions {
+            selected_unspent_transactions.push(unspent_transaction);
+            transferred_amount =
+                (transferred_amount + value).context(ErrorKind::BalanceAdditionError)?;
+
+            if transferred_amount >= amount_to_transfer {
+                break;
+            }
+        }
+
+        if transferred_amount < amount_to_transfer {
+            Err(ErrorKind::InsufficientBalance.into())
+        } else {
+            Ok((
+                selected_unspent_transactions,
+                (transferred_amount - amount_to_transfer)
+                    .context(ErrorKind::BalanceAdditionError)?,
+            ))
+        }
+    }
+
+    fn witnesses<W: WalletClient>(
+        &self,
+        name: &str,
+        passphrase: &SecStr,
+        wallet_client: &W,
+        transaction: &Tx,
+    ) -> Result<TxWitness> {
+        let mut witnesses = Vec::with_capacity(transaction.inputs.len());
+
+        for input in &transaction.inputs {
+            let input = wallet_client.output(&input.id, input.index)?;
+
+            match wallet_client.private_key(name, passphrase, &input.address)? {
+                None => return Err(ErrorKind::PrivateKeyNotFound.into()),
+                Some(private_key) => witnesses.push(private_key.sign(&transaction.id())?),
+            }
+        }
+
+        Ok(witnesses.into())
+    }
+
+    fn build_with_fee<W: WalletClient>(
+        &self,
+        name: &str,
+        passphrase: &SecStr,
+        mut outputs: Vec<TxOut>,
+        attributes: TxAttributes,
+        wallet_client: &W,
+        fee: Fee,
+    ) -> Result<TxAux> {
+        let unspent_transactions = wallet_client.unspent_transactions(name, passphrase)?;
```

Comment:
does this return all unspent transactions? if so, why not to pass them as an argument instead of retrieving them and re-sorting them (`build_with_fee` is called in a while loop below https://github.com/crypto-com/chain/pull/80/files#diff-d3adbcf3f20e967015c575da2cd85dfdR158)?